virttest/qemu_devices/qcontainer.py

```diff
@@ -1903,3 +1903,51 @@ def memory_define_by_params(self, params, name):
                 dimm.set_param("memdev", mem.get_qid())
                 devices.append(dimm)
         return devices
+
+    def input_define_by_params(self, params, name, bus=None):
+        """
+        Create input device by params.
+
+        :param params: VM params.
+        :param name: Object name of input device.
+        :param bus: Parent bus.
+        """
+        params = params.object_params(name)
+        dev_map = {"mouse": {"virtio": "virtio-mouse"},
+                   "keyboard": {"virtio": "virtio-keyboard"},
+                   "tablet": {"virtio": "virtio-tablet"}}
+        dev_type = params["input_dev_type"]
+        bus_type = params["input_dev_bus_type"]
+        driver = dev_map.get(dev_type)
+        if not driver:
+            raise ValueError("unsupported input device type")
+        driver = driver.get(bus_type)
+        if not driver:
+            raise ValueError("unsupported input device bus")
+
+        machine_type = params.get("machine_type", "")
+        qbus_type = "PCI"
+        if machine_type.startswith("q35") or machine_type.startswith("arm64"):
+            qbus_type = "PCIE"
+
+        if bus_type == "virtio":
+            if "-mmio:" in machine_type:
+                driver += "-device"
+                qbus_type = "virtio-bus"
+            elif machine_type.startswith("s390"):
+                driver += "-ccw"
+                qbus_type = "virtio-bus"
+            else:
+                driver += "-pci"
```

Comment:
Looks like we also need add suffix for many other virtio devices, do we consider make a general function for it? Something like:
```
         def get_virtio_device_and_bus(param, driver):
             machine_type = params.get("machine_type", "")
             qbus_type = "PCI"
             if machine_type.startswith("q35") or machine_type.startswith("arm64"):
                 qbus_type = "PCIE"
￼             if "-mmio:" in machine_type:
￼                 driver += "-device"
￼                 qbus_type = "virtio-bus"
￼             elif machine_type.startswith("s390"):
￼                 driver += "-ccw"
￼                 qbus_type = "virtio-bus"
￼             else:
￼                 driver += "-pci"
             return driver, bus_type
````

Reply:
Or configure it in machine.cfg? Like: https://github.com/avocado-framework/avocado-vt/pull/1639/files#diff-21bb94094179a0d14ffbd30ae8357b3fR56

Reply:
A more sensible way is to implement these with OOP, it will go step-by-step in the future since requiring lots of restructure works to be done.

Reply:
@luckyh I see, do you think my modification on machine.cfg here is necessary in this stage, should I drop it? https://github.com/avocado-framework/avocado-vt/pull/1639/files#diff-21bb94094179a0d14ffbd30ae8357b3fR56

Reply:
I don't suggest to make it user-configurable, so I think we can drop it, thanks.

Reply:
@luckyh Okay, I will update it, thanks.